### ISSUE TITLE
Bug: X loader regression introduced with commit f844c33 on 3 Nov 2023

### ISSUE THREAD

**Issue by tellypresence** [assimp/assimp#5332]
Appears to be a recent regression in X loader

Loading bundled model [`test/models-nonbsd/X/dwarf.x`](https://github.com/assimp/assimp/blob/master/test/models-nonbsd/X/dwarf.x) exhibits strange artifacts unless `aiProcess_PreTransformVertices` flag is applied during load (which of course destroys rigging/animation)

Problems observed when loaded without `aiProcess_PreTransformVertices` flag
- Other than (most of) the head, helmet and left shoulder pad, the body mesh is not visible.
- Animation for the body mesh seems to be correct
- The axe is fully visible, but no longer animates (used to move with the body)

Note that this problem can also be observed in bundled IRR scenes which load the dwarf perserving bones/animation

<table>
<tr><th>Screenshots</th><th>Notes</th></tr>
<tr><td>
<img src="https://github.com/assimp/assimp/assets/26958898/61765fad-2806-4f39-9b37-097221c2d793" alt="x_loader_regression_normal_load_bones_n_anim_preserved_Screenshot_20231118-092548" width=120 />
<img src="https://github.com/assimp/assimp/assets/26958898/5003ba97-40d7-4d70-87b6-8f8be25ed3b0" alt="x_loader_regression_normal_load_bones_n_anim_preserved_Screenshot_20231118-092557" width=120 />
<img src="https://github.com/assimp/assimp/assets/26958898/079c9c5b-464a-4aa2-b7ce-6e6dac90be30" alt="x_loader_regression_normal_load_bones_n_anim_preserved_az_recorder_20231118_112405" width=120 />

</td><td>Loaded without

<tt>aiProcess_PreTransformVertices</tt> flag 
(bones/animation preserved)

Notes:
 - The axe is no longer animated
 - double sided rendering was not 
enabled here, but the left shoulder 
pad can be viewed from both top
and bottom</td></tr>
<tr><td>
<img src="https://github.com/assimp/assimp/assets/26958898/dfd5ebcd-accf-4656-8d22-e05fc5b89a9a" alt="x_loader_regression_pretransform_verts_bones_n_anim_lost_Screenshot_20231118-093548" width=120 />
</td><td>Loaded with

<tt>aiProcess_PreTransformVertices</tt> flag
(bones/animation lost)</td></tr>
</table>

### Bisect
Bug was introduced on 3 Nov 2023; the X loader .cpp/.h files were touched here.

<table>
<tr><th>Assimp commit</th><th>Screenshots</th><th>Notes</th></tr>
<tr><td><a href="https://github.com/assimp/assimp/commit/f844c3397d7726477ab0fdca8efd3df56c18366b"><tt>f844c33</tt></a><br>3 Nov 2023</td><td>
<img src="https://github.com/assimp/assimp/assets/26958898/079c9c5b-464a-4aa2-b7ce-6e6dac90be30" alt="x_loader_regression_normal_load_bones_n_anim_preserved_az_recorder_20231118_112405" width=120 />
</td><td>Bug introduced here</td></tr>
<tr><td><a href="https://github.com/assimp/assimp/commit/a7cfa3264a9ab2b6694fc79f09b14494921b2e04"><tt>a7cfa32</tt></a><br>2 Nov 2023</td><td>
<img src="https://github.com/assimp/assimp/assets/26958898/3f5232a8-4fa4-446d-a396-59f4830144d4" alt="x_loader_was_ok_az_recorder_20231118_160835" width=120 />
</td><td>OK</td></tr>
</table>

**Comment by tellypresence** [comment]
### TL;DR
New `boneIdx` and conditional around line ~367 causes the obvious breakage
```
const unsigned int boneIdx = obone.mWeights[d].mVertex; // <-- new line
if (boneIdx < obone.mWeights.size()) { // <-- new line
    oldWeights[obone.mWeights[d].mVertex] = obone.mWeights[d].mWeight; // <-- original line
} // <-- new line
```

### Additional analysis details
AFAICT there are only 3 non-whitespace changes of note in the offending commit:

Line ~315 changed: checked "Before" vs "After" in isolation, both appear to work OK
```
Before
if ( newIdx > sourceMesh->mPositions.size() ) {

After
if (newIdx >= sourceMesh->mPositions.size()) {
```

New conditional block added around line ~330: checked the conditional in isolation, both with and without appear to work OK
```
if (idx < sourceMesh->mNormals.size()) {
    mesh->mNormals[newIndex] = sourceMesh->mNormals[idx]; // <-- original line
}
```

New `boneIdx` and conditional around line ~367: checked the conditional in isolation, this causes the obvious breakage
```
const unsigned int boneIdx = obone.mWeights[d].mVertex;
if (boneIdx < obone.mWeights.size()) {
    oldWeights[obone.mWeights[d].mVertex] = obone.mWeights[d].mWeight; // <-- original line
}
```

**Comment by tellypresence** [comment]
PR #5372 will fix this bug once merged

**Comment by tellypresence** [comment]
#5372 merged, should be fixed
